prealloc.go

```diff
@@ -198,13 +198,31 @@ func exists(filename string) bool {
 	return err == nil
 }
 
+func contains(slice []string, item string) bool {
+	for _, s := range slice {
+		if s == item {
+			return true
+		}
+	}
+
+	return false
+}
+
+var arrayTypes []string
```

Comment:
I think this might cause trouble - `arrayTypes` will never get cleared, correct? I did a little testing on the 1.9.2 standard library and this slice grows pretty rapidly if it's being run on the entire thing: 
```
block headerV7 headerGNU headerSTAR headerUSTAR sparseArray sparseNode readBuf writeBuf scriptedReader asciiSet MultiFlag SparseTree eqclass LocPair byKey ArchsByName byPath byFileName blocksByStart boundariesByPos blockSlice byName byDate StringsFlag ImportStack actionQueue VarsByOffset Opcross ocmp ocmp ocmp ocmp ocmp uint64s byAddr machoscmp byExtname byTypeStr byPkg bySizeAndName rbyoff gtypeList gdescList typePath moveToFrontDecoder huffmanSymbolLengthPairs huffmanCodes byLiteral byFreq IntHeap PriorityQueue myHeap gcmCount gcmHashKey p224FieldElement p224LargeFieldElement RDNSequence RelativeDistinguishedNameSET userDefinedSlice userDefinedBytes userDefinedBytes RawBytes bs is LoadBytes StringTable ObjectIdentifier RawContent testSET bytesEncoder multiEncoder oidEncoder RDNSequence RelativeDistinguishedNameSET RawMessage byteKind byteKind renamedByteSlice renamedRenamedByteSlice byIndex T T CharData Comment Directive MyBytes interval flagVar hexBytes renamedBytes SE buffer byImportSpec byCommentPos byPos byInterval nodeStack byPath ErrorList Table sum128 sum128a Palette
```

It seems like it'd be a pain to maintain the list and purge unexported typedefs after we get into another package, but I did have an idea:

Up in `checkForPreallocations`, we're ranging through our files:
```Go
    for _, f := range files {                                                                
          ast.Walk(retVis, f)                                                                  
      } 
```

My thought was we could add an additional field to `returnsVisitor` (we could probably still call it `arrayTypes`) and keep our existing logic in `Visit`, except we'd clear our arrayTypes field after each call to `Walk`:

```Go
    for _, f := range files {                                                          
          retVis.arrayTypes = nil
          ast.Walk(retVis, f)                                                                  
      } 
```

I think this would help bound the size of `arrayTypes` (while I think still supporting package scoped typedefs within the same package). Thoughts? Sorry to drag this review on, but I think we're getting close!

Reply:
To test `arrayTypes` growing, I just adding a print statement prior to returning in `Visit`:

```Go
if len(arrayTypes) > 0 {
	file := v.f.File(node.Pos())
	fmt.Printf("%v %v\n", file.Name(), arrayTypes)
}
return v
```

Reply:
sure. Yes i agree. actually my bad. i haven't been paying much attention to this fix as i should have had. as a workaround to purging the typedefs list, what you suggested makes sense. 👍 I've made the changes, can you run it once more on the same codebase with the above changes.